gateway/webhook.go

```diff
@@ -0,0 +1,308 @@
+	webhook.gwHandle.trackRequestMetrics(resp.err)
+	if resp.err != "" {
+		code := 400
+		if resp.statusCode != 0 {
+			code = resp.statusCode
+		}
+		logger.Debug(resp.err)
+		http.Error(w, resp.err, code)
+	} else {
+		logger.Debug(getStatus(Ok))
+		w.Write([]byte(getStatus(Ok)))
+	}
+}
+
+func (webhook *webhookHandleT) batcher(sourceDef string) {
+	var reqBuffer = make([]*webhookT, 0)
+	timeout := time.After(webhookBatchTimeout)
+	for {
+		select {
+		case req := <-webhook.requestQ[sourceDef]:
+			//Append to request buffer
+			reqBuffer = append(reqBuffer, req)
+			if len(reqBuffer) == maxWebhookBatchSize {
+				breq := batchWebhookT{batchRequest: reqBuffer, sourceType: sourceDef}
+				webhook.batchRequestQ <- &breq
+				reqBuffer = nil
+				reqBuffer = make([]*webhookT, 0)
+			}
+		case <-timeout:
+			timeout = time.After(webhookBatchTimeout)
+			if len(reqBuffer) > 0 {
+				breq := batchWebhookT{batchRequest: reqBuffer, sourceType: sourceDef}
+				webhook.batchRequestQ <- &breq
+				reqBuffer = nil
+				reqBuffer = make([]*webhookT, 0)
+			}
+		}
+	}
+}
+
+func (bt *batchWebhookTransformerT) batchTransform() {
+	for breq := range bt.webhook.batchRequestQ {
+		payloadArr := [][]byte{}
+		webRequests := []*webhookT{}
+		for _, req := range breq.batchRequest {
+			body, err := ioutil.ReadAll(req.request.Body)
+			req.request.Body.Close()
+
+			if err != nil {
+				req.done <- webhookErrorRespT{err: getStatus(RequestBodyReadFailed)}
+				continue
+			}
+
+			payloadArr = append(payloadArr, body)
+			webRequests = append(webRequests, req)
+		}
+
+		if len(payloadArr) == 0 {
+			continue
+		}
+
+		if _, ok := bt.stats.sourceStats[breq.sourceType]; !ok {
+			bt.stats.sourceStats[breq.sourceType] = newWebhookStat(breq.sourceType)
+		}
+
+		// stats
+		bt.stats.sourceStats[breq.sourceType].numEvents.Count(len(payloadArr))
+		bt.stats.sourceStats[breq.sourceType].sourceTransform.Start()
+
+		batchResponse := bt.transform(payloadArr, breq.sourceType)
+
+		// stats
+		bt.stats.sourceStats[breq.sourceType].sourceTransform.End()
+
+		if batchResponse.batchError != nil {
+			statusCode := 500
+			if batchResponse.statusCode != 0 {
+				statusCode = batchResponse.statusCode
+			}
+			for _, req := range breq.batchRequest {
+				req.done <- webhookErrorRespT{statusCode: statusCode, err: batchResponse.batchError.Error()}
+			}
+			continue
+		}
+
+		if len(batchResponse.responses) != len(payloadArr) {
+			panic("webhook batchtransform response events size does not equal sent events size")
+		}
+
+		bt.stats.sourceStats[breq.sourceType].numOutputEvents.Count(len(batchResponse.responses))
+
+		for idx, resp := range batchResponse.responses {
+			webRequest := webRequests[idx]
+			output := resp.output
+			if resp.err != "" {
+				webRequests[idx].done <- webhookErrorRespT{err: resp.err, statusCode: resp.statusCode}
+				continue
+			}
+			rruntime.Go(func() {
+				bt.webhook.enqueueToWebRequestQ(webRequest, output)
+			})
```

Comment:
Isn't it better to have some workers to do this enqueing?
Assume the following random case: If we have 1000 reqs per sec to /v1/webhook. They would translate to 1000 batchresponses which in turn would spawn 1000 goroutines. 

Ignore what I said dont make sense.